ui/components/TracePage/TracePage.tsx

```diff
@@ -180,11 +209,58 @@ const TreeProvider = ({ traceId, children }: { traceId: string; children: ReactN
     return (id: string) => focussedIds.includes(id);
   }, [selectedId, calls, rootId]);
 
+  // Placeholders for this render cycle to avoid fetching the same block multiple times.
+  // This gets updated during rendering, which is why we don't use state.
+  const blockRequests = useRef<Set<number>>(new Set());
+
+  function getBlockValue<T>(blockAddress: BlockAddress<T>): T | undefined {
+    const [blockNumber, blockLineno] = blockAddress;
+    const block = blocks[blockNumber];
+    if (block) {
+      if (blockLineno < block.length) {
+        return JSON.parse(block[blockLineno]);
+      }
+      return undefined; // wait for the other lines
+    }
+
+    if (blockRequests.current.has(blockNumber)) {
+      // This block has already been requested.
+      return undefined;
+    }
+    blockRequests.current.add(blockNumber);
+
+    const url = `${urlPrefix(traceId)}/block_${blockNumber}.jsonl`;
+    const fetchBlock = async (start: number) => {
+      const response = await fetch(url, { headers: { Range: `bytes=${start}-999999999` } });
```

Comment:
You can leave off the 999999999 and it's still a valid request (but `routes.traces.get_file` will need to be updated to parse that correctly)